### PR TITLE
Add V1beta1CSRAPICompatibility feature gate for registration-controller

### PR DESCRIPTION
**Pull request by promacanthus**
In some cases, our hub cluster's kubernetes version is less than **1.19**, and its `certificates.k8s.io` is **v1beta1**. So we need to enable the feature gate V1beta1CSRAPICompatibility. Otherwise, the registration-controller will get the following error message.

```shell
W0630 07:07:55.361239       1 reflector.go:324] k8s.io/client-go@v0.23.0/tools/cache/reflector.go:167: failed to list *v1.CertificateSigningRequest: the server could not find the requested resource
E0630 07:07:55.361260       1 reflector.go:138] k8s.io/client-go@v0.23.0/tools/cache/reflector.go:167: Failed to watch *v1.CertificateSigningRequest: failed to list *v1.CertificateSigningRequest: the server could not find the requested resource
```

Signed-off-by: Promacanthus <promacanthus@gmail.com>

### CODE REVIEW

File: pkg/hub/manager.go [hunk 1]
```diff
@@ -82,11 +83,20 @@ func RunControllerManager(ctx context.Context, controllerContext *controllercmd.
 		controllerContext.EventRecorder,
 	)
 
-	csrController := csr.NewCSRApprovingController(
-		kubeClient,
-		kubeInfomers.Certificates().V1().CertificateSigningRequests(),
-		controllerContext.EventRecorder,
-	)
+	var csrController factory.Controller
+	if features.DefaultHubMutableFeatureGate.Enabled(ocmfeature.V1beta1CSRAPICompatibility) {
```
Comment (skeeey) [inline pkg/hub/manager.go:90]: I think we need more judgement at here, like:  
https://github.com/open-cluster-management-io/registration/blob/main/pkg/clientcert/cert_controller.go#L140

File: pkg/hub/csr/controller_beta.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		csrLister:     v1beta1CSRInformer.Lister(),
+		eventRecorder: recorder.WithComponentSuffix("csr-approving-controller"),
+	}
+
+	return factory.New().WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+		accessor, _ := meta.Accessor(obj)
+		return accessor.GetName()
+	}, v1beta1CSRInformer.Informer()).
+		WithSync(c.sync).
+		ToController("CSRApprovingController", recorder)
```
Comment (skeeey) [inline pkg/hub/csr/controller_beta.go]: V1Beta1CSRApprovingController

File: pkg/hub/csr/controller_beta.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		return err
+	}
+	c.eventRecorder.Eventf("ManagedClusterCSRAutoApproved", "spoke cluster csr %q is auto approved by hub csr controller", csr.Name)
+	return nil
+}
+
+// To check a renewal managed cluster csr, we check
+// 1. if the signer name in csr request is valid.
+// 2. if organization field and commonName field in csr request is valid.
+// 3. if user name in csr is the same as commonName field in csr request.
```
Comment (skeeey) [inline pkg/hub/csr/controller_beta.go]: I think this comments should be for `isV1beta1SpokeClusterClientCertRenewal `

File: pkg/hub/csr/controller_beta.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		},
+	}
+	sar, err := c.kubeClient.AuthorizationV1().SubjectAccessReviews().Create(ctx, sar, metav1.CreateOptions{})
+	if err != nil {
+		return false, err
+	}
+	return sar.Status.Allowed, nil
+}
+
+func isV1beta1SpokeClusterClientCertRenewal(csr *certificatesv1beta1.CertificateSigningRequest) bool {
```
Comment (skeeey) [inline pkg/hub/csr/controller_beta.go]: this logic of this function and `authorize` function are same with the csr v1, how about we define a common struct to store the csr info, and pass this struct to these function.

```golang
type csrInfo struct {
	name       string
	singerName string
	userName   string
	uid        string
	groups     []string
	extra      map[string]authorizationv1.ExtraValue
	request    []byte
}

func authorize(ctx context.Context, csr csrInfo) (bool, error)
func isSpokeClusterClientCertRenewal(csr csrInfo) bool
```

**Comment by skeeey** [comment]
@Promacanthus thanks for your contribution, I think you need run `go mod tidy` and `go mod vendor` to make the `verify deps` pass

**Comment by promacanthus** [comment]
> @Promacanthus thanks for your contribution, I think you need run `go mod tidy` and `go mod vendor` to make the `verify deps` pass

Thanks for your review. I will rerun the `go mod tidy` and `go mod vendor` to fix the `verify-deps` problem and follow your suggestions above.

**Comment by promacanthus** [comment]
@skeeey I have modify the code based on your comment, please review it again when you have time.

I think  this [PR](https://github.com/open-cluster-management-io/api/pull/169) needs to be merged first. 

When I rerun the `go mod vendor` command, there is no `V1beta1CSRAPICompatibility` in `DefaultHubRegistrationFeatureGates` at `vendor/open-cluster-management.io/api/feature/feature.go` file. 

This will break integration and e2e test, and get the error message as following.

```shell
feature "V1beta1CSRAPICompatibility" is not registered in FeatureGate 

"open-cluster-management.io/registration/pkg/features/feature.go:15"
```

**Comment by skeeey** [comment]
@Promacanthus the integration test failed

```
panic: feature "V1beta1CSRAPICompatibility" is not registered in FeatureGate "open-cluster-management.io/registration/pkg/features/feature.go:15"
```

you need register the feature gate to the [`management.io/registration/pkg/features/feature.go`](https://github.com/open-cluster-management-io/registration/blob/main/pkg/features/feature.go#L18)

File: pkg/hub/csr/controller.go [hunk 5]
```diff
@@ -105,18 +106,22 @@ func (c *csrApprovingController) sync(ctx context.Context, syncCtx factory.SyncC
 
 // Using SubjectAccessReview API to check whether a spoke agent has been authorized to renew its csr,
 // a spoke agent is authorized after its spoke cluster is accepted by hub cluster admin.
-func (c *csrApprovingController) authorize(ctx context.Context, csr *certificatesv1.CertificateSigningRequest) (bool, error) {
-	extra := make(map[string]authorizationv1.ExtraValue)
-	for k, v := range csr.Spec.Extra {
-		extra[k] = authorizationv1.ExtraValue(v)
+func (c *csrApprovingController) authorize(ctx context.Context, csr csrInfo) (bool, error) {
```
Comment (skeeey) [inline pkg/hub/csr/controller.go]: @Promacanthus would also refactor this function to

```
// Using SubjectAccessReview API to check whether a spoke agent has been authorized to renew its csr,
// a spoke agent is authorized after its spoke cluster is accepted by hub cluster admin.
func authorize(ctx context.Context, kubeClient kubernetes.Interface, csr  *crsInfo) (bool, error) {
...
}
```

thus both v1 and v1beta1 can use it to authroize

**Comment by promacanthus** [comment]
> @Promacanthus the integration test failed
> 
> ```
> panic: feature "V1beta1CSRAPICompatibility" is not registered in FeatureGate "open-cluster-management.io/registration/pkg/features/feature.go:15"
> ```
> 
> you need register the feature gate to the [`management.io/registration/pkg/features/feature.go`](https://github.com/open-cluster-management-io/registration/blob/main/pkg/features/feature.go#L18)

The previous understanding of `DefaultHubRegistrationFeatureGates` is not correct, I will close the [PR](https://github.com/open-cluster-management-io/api/pull/169) and register the `V1beta1CSRAPICompatibility` feature gate to `pkg/features/feature.go`.

File: pkg/hub/csr/controller.go (hunk 5 above)
Comment (promacanthus) [reply pkg/hub/csr/controller.go]: I will refactor the `authorize()` function.

**Comment by skeeey** [comment]
> > @Promacanthus the integration test failed
> > ```
> > panic: feature "V1beta1CSRAPICompatibility" is not registered in FeatureGate "open-cluster-management.io/registration/pkg/features/feature.go:15"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > you need register the feature gate to the [`management.io/registration/pkg/features/feature.go`](https://github.com/open-cluster-management-io/registration/blob/main/pkg/features/feature.go#L18)
> 
> The previous understanding of `DefaultHubRegistrationFeatureGates` is not correct, I will close the [PR](https://github.com/open-cluster-management-io/api/pull/169) and register the `V1beta1CSRAPICompatibility` feature gate to `pkg/features/feature.go`.

Hi, @Promacanthus I think for this case, it make sense to add the `V1beta1CSRAPICompatibility` to the `DefaultHubRegistrationFeatureGates` in the api repo, so you do not need the close the pr, we can wait for the api pr is merged,  then we update the api dependency in this pr to use the new `DefaultHubRegistrationFeatureGates`, what's your opinion?

**Comment by promacanthus** [comment]
> > > @Promacanthus the integration test failed
> > > ```
> > > panic: feature "V1beta1CSRAPICompatibility" is not registered in FeatureGate "open-cluster-management.io/registration/pkg/features/feature.go:15"
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > you need register the feature gate to the [`management.io/registration/pkg/features/feature.go`](https://github.com/open-cluster-management-io/registration/blob/main/pkg/features/feature.go#L18)
> > 
> > 
> > The previous understanding of `DefaultHubRegistrationFeatureGates` is not correct, I will close the [PR](https://github.com/open-cluster-management-io/api/pull/169) and register the `V1beta1CSRAPICompatibility` feature gate to `pkg/features/feature.go`.
> 
> Hi, @Promacanthus I think for this case, it make sense to add the `V1beta1CSRAPICompatibility` to the `DefaultHubRegistrationFeatureGates` in the api repo, so you do not need the close the pr, we can wait for the api pr is merged, then we update the api dependency in this pr to use the new `DefaultHubRegistrationFeatureGates`, what's your opinion?

Agreed with you. Let me reopen the PR.

**Comment by skeeey** [comment]
This pr is fine to me

@qiujian16 would you also take a look this? if there are no more comments, I think we can merge this

**Comment by qiujian16** [comment]
/lgtm 
Thanks!

**Pull request merged**